jdbc/src/java/org/apache/hive/jdbc/saml/HiveJdbcBrowserClient.java

```diff
@@ -0,0 +1,322 @@
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.io.UnsupportedEncodingException;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketTimeoutException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.apache.hive.jdbc.Utils.JdbcConnectionParams;
+import org.apache.hive.service.auth.saml.HiveSamlUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used to execute a browser based SSO workflow with the authentication mode
+ * is browser.
+ */
+public class HiveJdbcBrowserClient implements IJdbcBrowserClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HiveJdbcBrowserClient.class);
+  // error message when the socket times out.
+  @VisibleForTesting
+  public static final String TIMEOUT_ERROR_MSG = "Timed out while waiting for server response";
+  private final ServerSocket serverSocket;
+  private HiveJdbcBrowserServerResponse serverResponse;
+  protected JdbcBrowserClientContext clientContext;
+  // By default we wait for 2 min unless overridden by a JDBC connection param
+  // browserResponseTimeout
+  private static final int DEFAULT_SOCKET_TIMEOUT_SECS = 120;
+  private final ExecutorService serverResponseThread = Executors.newSingleThreadExecutor(
+      new ThreadFactoryBuilder().setNameFormat("Hive-Jdbc-Browser-Client-%d")
+          .setDaemon(true).build());
+
+  HiveJdbcBrowserClient(JdbcConnectionParams connectionParams)
+      throws HiveJdbcBrowserException {
+    serverSocket = getServerSocket(connectionParams.getSessionVars());
+  }
+
+  private ServerSocket getServerSocket(Map<String, String> sessionConf)
+      throws HiveJdbcBrowserException {
+    final ServerSocket serverSocket;
+    int port = Integer.parseInt(sessionConf
+        .getOrDefault(JdbcConnectionParams.AUTH_BROWSER_RESPONSE_PORT, "0"));
+    int timeout = Integer.parseInt(
+        sessionConf.getOrDefault(JdbcConnectionParams.AUTH_BROWSER_RESPONSE_TIMEOUT_SECS,
+            String.valueOf(DEFAULT_SOCKET_TIMEOUT_SECS)));
+    try {
+      serverSocket = new ServerSocket(port, 0,
+          InetAddress.getByName(HiveSamlUtils.LOOP_BACK_INTERFACE));
+      LOG.debug("Browser response timeout is set to {} seconds", timeout);
+      serverSocket.setSoTimeout(timeout * 1000);
+    } catch (IOException e) {
+      throw new HiveJdbcBrowserException("Unable to bind to the localhost");
+    }
+    return serverSocket;
+  }
+
+  public Integer getPort() {
+    return serverSocket.getLocalPort();
```

Comment:
should this account for serverSocket being null in case of an exception in getServerSocket() ?

Reply:
serverSocket is a final field and it will always be not-null. If there is an exception during getServerSocket, the HiveJdbcBrowserClient's constructor will error out.